File: app/lib/services/audio_service.dart
import 'dart:developer' as developer;
import 'package:audioplayers/audioplayers.dart';

enum SfxType {
  buttonClick('assets/audio/sfx/button_click.wav'),
  pageTurn('assets/audio/sfx/page_turn.wav'),
  diceRoll('assets/audio/sfx/dice_roll.wav'),
  diceSuccess('assets/audio/sfx/dice_settle_success.wav'),
  diceFail('assets/audio/sfx/dice_settle_fail.wav'),
  potionDrink('assets/audio/sfx/potion_drink.wav'),
  equipGear('assets/audio/sfx/equip_gear.wav');

  final String assetPath;
  const SfxType(this.assetPath);
}

enum AmbientTrack {
  citadelWind('assets/audio/ambient/ambient_citadel_wind.wav'),
  dungeonDrips('assets/audio/ambient/ambient_dungeon_drips.wav'),
  cryptDrone('assets/audio/ambient/ambient_crypt_drone.wav'),
  none('');

  final String assetPath;
  const AmbientTrack(this.assetPath);

  static AmbientTrack fromLocation(String? locationId) {
    if (locationId == null || locationId.isEmpty) return AmbientTrack.citadelWind;
    final loc = locationId.toLowerCase();
    if (loc.contains('dungeon') || loc.contains('cell') || loc.contains('catacomb') || loc.contains('prison')) {
      return AmbientTrack.dungeonDrips;
    }
    if (loc.contains('crypt') || loc.contains('altar') || loc.contains('vault') || loc.contains('arcane')) {
      return AmbientTrack.cryptDrone;
    }
    return AmbientTrack.citadelWind;
  }
}

class AudioService {
  static final AudioService _instance = AudioService._internal();
  factory AudioService() => _instance;
  AudioService._internal();

  final AudioPlayer _sfxPlayer = AudioPlayer();
  final AudioPlayer _ambientPlayer = AudioPlayer();

  bool _isSfxMuted = false;
  bool _isAmbientMuted = false;
  double _sfxVolume = 0.85;
  double _ambientVolume = 0.40;
  AmbientTrack _currentAmbient = AmbientTrack.none;

  bool get isSfxMuted => _isSfxMuted;
  bool get isAmbientMuted => _isAmbientMuted;
  double get sfxVolume => _sfxVolume;
  double get ambientVolume => _ambientVolume;
  AmbientTrack get currentAmbient => _currentAmbient;

  Future<void> init() async {
    try {
      await _ambientPlayer.setReleaseMode(ReleaseMode.loop);
      await _ambientPlayer.setVolume(_isAmbientMuted ? 0.0 : _ambientVolume);
      await _sfxPlayer.setVolume(_isSfxMuted ? 0.0 : _sfxVolume);
    } catch (e) {
      developer.log('AudioService init warning (expected in tests/unsupported audio devices): $e');
    }
  }

  Future<void> playSfx(SfxType sfx) async {
    if (_isSfxMuted) return;
    try {
      await _sfxPlayer.stop();
      await _sfxPlayer.setVolume(_sfxVolume);
      await _sfxPlayer.play(AssetSource(sfx.assetPath.replaceFirst('assets/', '')));
    } catch (e) {
      developer.log('Failed to play SFX: ${sfx.assetPath}, error: $e');
    }
  }

  Future<void> playAmbient(AmbientTrack track) async {
    if (_currentAmbient == track) return;
    _currentAmbient = track;

    if (track == AmbientTrack.none || track.assetPath.isEmpty) {
      await stopAmbient();
      return;
    }

    if (_isAmbientMuted) return;

    try {
      await _ambientPlayer.stop();
      await _ambientPlayer.setReleaseMode(ReleaseMode.loop);
      await _ambientPlayer.setVolume(_ambientVolume);
      await _ambientPlayer.play(AssetSource(track.assetPath.replaceFirst('assets/', '')));
    } catch (e) {
      developer.log('Failed to play ambient track: ${track.assetPath}, error: $e');
    }
  }

  Future<void> stopAmbient() async {
    try {
      await _ambientPlayer.stop();
    } catch (e) {
      developer.log('Failed to stop ambient track: $e');
    }
  }

  Future<void> toggleSfxMute() async {
    _isSfxMuted = !_isSfxMuted;
    try {
      await _sfxPlayer.setVolume(_isSfxMuted ? 0.0 : _sfxVolume);
    } catch (_) {}
  }

  Future<void> toggleAmbientMute() async {
    _isAmbientMuted = !_isAmbientMuted;
    try {
      if (_isAmbientMuted) {
        await _ambientPlayer.setVolume(0.0);
      } else {
        await _ambientPlayer.setVolume(_ambientVolume);
        if (_currentAmbient != AmbientTrack.none) {
          await playAmbient(_currentAmbient);
        }
      }
    } catch (_) {}
  }

  Future<void> setAmbientVolume(double volume) async {
    _ambientVolume = volume.clamp(0.0, 1.0);
    if (!_isAmbientMuted) {
      try {
        await _ambientPlayer.setVolume(_ambientVolume);
      } catch (_) {}
    }
  }

  Future<void> setSfxVolume(double volume) async {
    _sfxVolume = volume.clamp(0.0, 1.0);
    if (!_isSfxMuted) {
      try {
        await _sfxPlayer.setVolume(_sfxVolume);
      } catch (_) {}
    }
  }

  void dispose() {
    _sfxPlayer.dispose();
    _ambientPlayer.dispose();
  }
}

File: web/src/lib/play/audioService.ts
'use client';

export type SfxType =
  | 'buttonClick'
  | 'pageTurn'
  | 'diceRoll'
  | 'diceSuccess'
  | 'diceFail'
  | 'potionDrink'
  | 'equipGear';

const SFX_PATHS: Record<SfxType, string> = {
  buttonClick: '/audio/sfx/button_click.wav',
  pageTurn: '/audio/sfx/page_turn.wav',
  diceRoll: '/audio/sfx/dice_roll.wav',
  diceSuccess: '/audio/sfx/dice_settle_success.wav',
  diceFail: '/audio/sfx/dice_settle_fail.wav',
  potionDrink: '/audio/sfx/potion_drink.wav',
  equipGear: '/audio/sfx/equip_gear.wav',
};

export type AmbientTrack = 'citadelWind' | 'dungeonDrips' | 'cryptDrone' | 'none';

const AMBIENT_PATHS: Record<Exclude<AmbientTrack, 'none'>, string> = {
  citadelWind: '/audio/ambient/ambient_citadel_wind.wav',
  dungeonDrips: '/audio/ambient/ambient_dungeon_drips.wav',
  cryptDrone: '/audio/ambient/ambient_crypt_drone.wav',
};

export function ambientFromLocation(locationId?: string | null): AmbientTrack {
  if (!locationId) return 'citadelWind';
  const loc = locationId.toLowerCase();
  if (loc.includes('dungeon') || loc.includes('cell') || loc.includes('catacomb') || loc.includes('prison')) {
    return 'dungeonDrips';
  }
  if (loc.includes('crypt') || loc.includes('altar') || loc.includes('vault') || loc.includes('arcane')) {
    return 'cryptDrone';
  }
  return 'citadelWind';
}

interface AudioState {
  sfxMuted: boolean;
  ambientMuted: boolean;
  sfxVolume: number;
  ambientVolume: number;
  currentAmbient: AmbientTrack;
}

const listeners = new Set<(s: AudioState) => void>();
let state: AudioState = {
  sfxMuted: false,
  ambientMuted: false,
  sfxVolume: 0.85,
  ambientVolume: 0.4,
  currentAmbient: 'none',
};

function emit() {
  for (const l of listeners) l(state);
}

function setState(patch: Partial<AudioState>) {
  state = { ...state, ...patch };
  emit();
}

let ambientAudio: HTMLAudioElement | null = null;

export const audioService = {
  getState(): AudioState {
    return state;
  },
  subscribe(fn: (s: AudioState) => void): () => void {
    listeners.add(fn);
    return () => listeners.delete(fn);
  },

  playSfx(type: SfxType) {
    if (state.sfxMuted || typeof window === 'undefined') return;
    try {
      const el = new Audio(SFX_PATHS[type]);
      el.volume = state.sfxVolume;
      void el.play().catch(() => {});
    } catch {
      /* audio unsupported */
    }
  },

  playAmbient(track: AmbientTrack) {
    if (typeof window === 'undefined') return;
    if (track === 'none' || track === state.currentAmbient) {
      if (track === 'none') this.stopAmbient();
      return;
    }
    setState({ currentAmbient: track });
    if (state.ambientMuted) return;
    try {
      if (!ambientAudio) {
        ambientAudio = new Audio();
        ambientAudio.loop = true;
      }
      ambientAudio.src = AMBIENT_PATHS[track as Exclude<AmbientTrack, 'none'>];
      ambientAudio.volume = state.ambientVolume;
      void ambientAudio.play().catch(() => {});
    } catch {
      /* audio unsupported */
    }
  },

  stopAmbient() {
    setState({ currentAmbient: 'none' });
    if (ambientAudio) {
      ambientAudio.pause();
      ambientAudio.currentTime = 0;
    }
  },

  toggleSfxMute() {
    setState({ sfxMuted: !state.sfxMuted });
  },

  toggleAmbientMute() {
    const next = !state.ambientMuted;
    setState({ ambientMuted: next });
    if (next) {
      if (ambientAudio) ambientAudio.volume = 0;
    } else {
      if (ambientAudio) ambientAudio.volume = state.ambientVolume;
      if (state.currentAmbient !== 'none') this.playAmbient(state.currentAmbient);
    }
  },

  setAmbientVolume(v: number) {
    const vol = Math.max(0, Math.min(1, v));
    setState({ ambientVolume: vol });
    if (ambientAudio && !state.ambientMuted) ambientAudio.volume = vol;
  },

  setSfxVolume(v: number) {
    setState({ sfxVolume: Math.max(0, Math.min(1, v)) });
  },
};
